File: frontend/src/services/catalog.ts
import { fetchJson } from './http';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8001';

export interface CatalogItem {
  id: number;
  type: 'hunt' | 'quest' | 'custom';
  name: string;
  location: string;
  level_min?: number;
  level_max?: number;
  vocation?: string;
  exp_per_hour?: number;
  profit_per_hour?: number;
  creatures?: string;
  quest_reward?: string;
  quest_requirements?: string;
  strategy?: string;
  notes?: string;
  difficulty?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CatalogCreate {
  type: 'hunt' | 'quest' | 'custom';
  name: string;
  location: string;
  level_min?: number;
  level_max?: number;
  vocation?: string;
  exp_per_hour?: number;
  profit_per_hour?: number;
  creatures?: string;
  quest_reward?: string;
  quest_requirements?: string;
  strategy?: string;
  notes?: string;
  difficulty?: string;
}

const getAuthHeaders = () => {
  const token = localStorage.getItem('token');
  return {
    'Content-Type': 'application/json',
    ...(token && { Authorization: `Bearer ${token}` }),
  };
};

export const catalogApi = {
  async getItems(type?: string, level_min?: number, level_max?: number, vocation?: string): Promise<CatalogItem[]> {
    const params = new URLSearchParams();
    if (type) params.append('type', type);
    if (level_min) params.append('level_min', level_min.toString());
    if (level_max) params.append('level_max', level_max.toString());
    if (vocation) params.append('vocation', vocation);

    return fetchJson<CatalogItem[]>(`${API_URL}/catalog?${params}`, {
      headers: getAuthHeaders(),
    });
  },

  async getItem(id: number): Promise<CatalogItem> {
    return fetchJson<CatalogItem>(`${API_URL}/catalog/${id}`, {
      headers: getAuthHeaders(),
    });
  },

  async createItem(item: CatalogCreate): Promise<CatalogItem> {
    return fetchJson<CatalogItem>(`${API_URL}/catalog`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify(item),
      timeoutMode: 'admin',
    });
  },

  async updateItem(id: number, item: Partial<CatalogCreate>): Promise<CatalogItem> {
    return fetchJson<CatalogItem>(`${API_URL}/catalog/${id}`, {
      method: 'PUT',
      headers: getAuthHeaders(),
      body: JSON.stringify(item),
      timeoutMode: 'admin',
    });
  },

  async deleteItem(id: number): Promise<void> {
    await fetchJson(`${API_URL}/catalog/${id}`, {
      method: 'DELETE',
      headers: getAuthHeaders(),
      timeoutMode: 'admin',
    });
  },
};

File: backend/app/models/catalog.py
"""
Catalog Model - General catalog for hunts, quests, and other content
"""
from sqlalchemy import Column, Integer, String, Text, DateTime, Boolean, func
from app.db.database import Base

class Catalog(Base):
    __tablename__ = "catalog"
    __table_args__ = {'extend_existing': True}

    id = Column(Integer, primary_key=True, index=True)
    type = Column(String(50), nullable=False)  # 'hunt', 'quest', 'custom'
    name = Column(String(200), nullable=False)
    location = Column(String(200), nullable=False)
    level_min = Column(Integer, nullable=True)
    level_max = Column(Integer, nullable=True)
    vocation = Column(Text, nullable=True)
    exp_per_hour = Column(Integer, nullable=True)
    profit_per_hour = Column(Integer, nullable=True)
    creatures = Column(Text, nullable=True)
    
    # Quest specific fields
    quest_reward = Column(Text, nullable=True)
    quest_requirements = Column(Text, nullable=True)
    
    # General fields
    strategy = Column(Text, nullable=True)
    notes = Column(Text, nullable=True)
    difficulty = Column(String(50), nullable=True)  # 'easy', 'medium', 'hard', 'extreme'
    
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
